### PR TITLE
Redo cargo-miri logic

### PR DESCRIPTION
**Pull request by RalfJung**
This rewrite the cargo-miri logic for running the requested crate(s) following what we outlined in https://github.com/rust-lang/miri/issues/739: `cargo miri run/test $FLAGS` (almost) literally invokes `cargo run/test $FLAGS` but with some environment variables set so that we can control what happens:
* `RUSTC_WRAPPER` is set so that we get invoked instead of `rustc`. We use that power to mess with the flags being used for the build (things to be interpreted by Miri use a different sysroot), and when we are detecting a binary crate meant to be run by Miri, we grab the info we care about and put it into a JSON file for later use.
* `CARGO_TARGET_$TARGET_RUNNER` is set so what we get invoked when cargo wants to run a binary. At that point we take that JSON info from before and use it to invoke Miri.

Overall this works great! We get all sorts of cargo magic for free, and do not even need `cargo-metadata` any more. There's one annoying point though, which I have not managed to entirely work around yet: this means we are doing a full build, not just a check-build. Given that our sysroot is MIR-only, I was surprised that this even worked, but still -- this means we are doing more work than we should. So I also added some patching of arguments to make sure `--emit` does not contain `link`, and then more patching was required of the `--extern` flags for the binary because those referenced the `.rlib` files but now only `.rmeta` exists, and that is still not fully working because cargo seems to expect those `.rmeta` files and now triggers a rebuild each time as those files are still missing. My current plan is to make our wrapper create some empty dummy files with the right names, but the amount of hacks we are stacking on top of each other here is getting worrysome.^^ @ehuss your input would be welcome on this issue.

Due to re-using cargo more literally, this also changes flag parsing to match `cargo`. So `-Zmiri` flags now have to be passed via an environment variable (Cc https://github.com/rust-lang/miri/issues/1416).

This PR is not ready yet, but the parts that are there I think can be reviewed already. TODO:
* [x] [Fix Windows](https://github.com/rust-lang/miri/pull/1540#issuecomment-688733741).
* [x] Figure out how we can do check-only builds without the rebuild problem above. ~~I am also worried about cases where `build.rs` script might detect check-only builds and then do less work; I think some crates in rustc are doing that and if this is a thing in the wider ecosystem we need to find a way to support this as well.~~ (postponed that until we have a concrete example)
* [x] Currently cargo runs doctests as well, and they are not run in Miri. We should at least figure out a way to not run them at all (resolving https://github.com/rust-lang/miri/issues/584 is left for a future PR).
* [x] For some time, detect the old way of passing `-Zmiri` flags and warn that this will need updating. For some simple situations we can probably make it still support the old way, but I plan to remove those hacks after a bit. This is just to give people and me time to go around and send PRs to all projects that use Miri on CI, and update their use of the flags.
* [x] Add a test for stdin handling (https://github.com/rust-lang/miri/issues/1505). This should work now but we should be sure.
* [x] Add a test for cargo env vars (https://github.com/rust-lang/miri/issues/1515).
* [x] Check if https://github.com/rust-lang/miri/issues/1516 is resolved.
* [x] Check if https://github.com/rust-lang/miri/issues/1001 and https://github.com/rust-lang/miri/issues/1514 are resolved.
* [x] Check if https://github.com/rust-lang/miri/issues/1312 is resolved (not sure if it is wort adding a test).
* [x] Check if https://github.com/rust-lang/miri/issues/1512 is resolved (not sure if it is wort adding a test).

Fixes https://github.com/rust-lang/miri/issues/700.
Fixes https://github.com/rust-lang/miri/issues/739.
Fixes https://github.com/rust-lang/miri/issues/1001.
Fixes https://github.com/rust-lang/miri/issues/1312 (without a test, as we run without cargo's stdin/stdout wrapping now, as the test for stdin confirms).
Fixes https://github.com/rust-lang/miri/issues/1416.
Fixes https://github.com/rust-lang/miri/issues/1505.
Fixes https://github.com/rust-lang/miri/issues/1512 (without a test, as cargo now does all that handling anyway, which various other tests confirm).
Fixes https://github.com/rust-lang/miri/issues/1514.
Fixes https://github.com/rust-lang/miri/issues/1516.

Cc @alecmocatta who reported many of the bugs above; would be great if you could help with the tests e.g. by providing some small examples I could try.
r? @oli-obk

### CODE REVIEW

File: cargo-miri/bin.rs [hunk 1]
```diff
@@ -495,24 +502,36 @@ fn phase_cargo_rustc(mut args: env::Args) {
...
 
     let file = File::open(binary)
         .unwrap_or_else(|_| show_error(format!("File {:?} not found, or cargo-miri invoked incorrectly", binary)));
     let file = BufReader::new(file);
     let info: CrateRunInfo = serde_json::from_reader(file)
         .unwrap_or_else(|_| show_error(format!("File {:?} does not contain valid JSON", binary)));
     // FIXME: remove the file.
+
+    let mut cmd = miri();
+    // Forward rustc arguments,with our sysroot.
```
Comment (oli-obk) [inline cargo-miri/bin.rs]:
```suggestion
    // Forward rustc arguments, with our sysroot.
```

Comment (RalfJung) [reply cargo-miri/bin.rs]: This already got changed in a later commit.

**Comment by RalfJung** [comment]
CI says

> fatal error: File "target/i686-pc-windows-msvc/debug/cargo-miri-test.exe" not found or `cargo-miri` invoked incorrectly; please only invoke this binary through `cargo miri`

Looks like Windows needs a special case.

**Comment by RalfJung** [comment]
@oli-obk so regarding that check-build situation, I wonder if there isn't a better way than to mess with the flags... the best way of course would be to actually convince cargo to do a "check-build version of `cargo test`", but that might not be possible (Cc @ehuss). But as another option, is there a way to adjust [this code](https://github.com/rust-lang/miri/blob/604a674ea37b302fd605df67be10a24ce94ad0a6/src/bin/miri.rs#L167) to tell rustc to not do any codegen? Like, swap in a no-op-codegen-backend or so?

**Comment by bjorn3** [comment]
> But as another option, is there a way to adjust this code to tell rustc to not do any codegen? Like, swap in a no-op-codegen-backend or so?

`rustc_driver` currently doesn't have an interface to switch codegen backends apart from passing `-Zcodegen-backend=/path/to/codegen_backend.so`. I was actually planing to open a PR for this.

**Comment by RalfJung** [comment]
> rustc_driver currently doesn't have an interface to switch codegen backends apart from passing -Zcodegen-backend=/path/to/codegen_backend.so. I was actually planing to open a PR for this.

If we could pass `-Zcodegen-backend=none` and make the driver recognize that value to plug in some dummy backend, that would also work for us here I think.

I assume this means rustc would still produce both a .rlib and .rmeta file, so there is some unnecessary disk I/O, but I also assume that MIR-only .rlib are not big enough to make this a problem.

**Comment by bjorn3** [comment]
> If we could pass -Zcodegen-backend=none and make the driver recognize that value to plug in some dummy backend, that would also work for us here I think.

I added `-Zcodegen-backend=metadata-only` when I initially worked on making rustc compile without LLVM, but I removed it later because it was unused.

**Comment by RalfJung** [comment]
Ah, that sounds like exactly what we'd need here. :) Could you point me to the relevant PRs? Will that code still work today?

(If you have some time and want to just re-add it yourself that would be much appreciated. :D I can also do it though, just might take a few days until I get around to it.)

**Comment by bjorn3** [comment]
It was removed in https://github.com/rust-lang/rust/pull/58847.

**Comment by RalfJung** [comment]
Uff, this will be annoying... to implement the `supported_target_features` query, I'll basically have to [copy these arrays](https://github.com/rust-lang/rust/blob/7bc0bf72545f5d6ac6d9639391da875d5e147d9a/compiler/rustc_codegen_llvm/src/llvm_util.rs#L143). And that is after copying the code from the `hotplug_codegen_backend` test.

Also I am getting errors like this:
```
error[E0463]: can't find crate for `std`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: could not compile `std`.
```
Not sure where that is coming from. But this is much less clean than I expected...

**Comment by RalfJung** [comment]
Oh, I forgot that we are also using Miri for "host" crates that need actual binaries. Those will need a proper codegen backend. That probably explains the "can't find crate for" issue.

The problem with the target features remains, though.

File: cargo-miri/bin.rs [hunk 2]
```diff
@@ -566,55 +453,158 @@ fn inside_cargo_rustc() {
...
+            .unwrap_or_else(|_| show_error(format!("Cannot write to {:?}", filename.display())));
+        return;
+    }
+
+    let mut cmd = miri();
+    let mut emit_link_hack = false;
+    // Arguments are treated very differently depending on whether this crate is
+    // for interpretation by Miri, or for use by a build script / proc macro.
     if target_crate {
+        // Forward arguments, butremove "link" from "--emit" to make this a check-only build.
```
Comment (bjorn3) [inline cargo-miri/bin.rs:504]:
```suggestion
        // Forward arguments, but remove "link" from "--emit" to make this a check-only build.
```

**Comment by bjorn3** [comment]
> Uff, this will be annoying... to implement the supported_target_features query, I'll basically have to copy these arrays. And that is after copying the code from the hotplug_codegen_backend test.

I think at least the stable target features should be moved out of `rustc_codegen_llvm`, as they will need to be listed by every codegen backend for `corearch` to compile, even when none of it's functions are used.

File: cargo-miri/bin.rs (hunk 2 above)
Comment (RalfJung) [reply cargo-miri/bin.rs]: Thanks!

**Comment by RalfJung** [comment]
> I think at least the stable target features should be moved out of rustc_codegen_llvm, as they will need to be listed by every codegen backend for corearch to compile, even when none of it's functions are used.

Yeah, compiling that part of libstd is also where the errors show up for Miri.

**Comment by RalfJung** [comment]
I did some more work on the hacky approach of messing with `--emit`, and now got it to work reasonably well -- cargo doesn't incorrectly rebuild things any more, and it seems to work for Windows targets, too. So IMO this is an okay solution for now, though longer-term I'd like to move to something more robust, like a NOP codegen backend.

File: cargo-miri/bin.rs [hunk 3]
```diff
@@ -378,173 +371,83 @@ path = "lib.rs"
...
+    // Set the runner for the current target to us as well, so we can interpret the binaries.
+    let runner_env_name = format!("CARGO_TARGET_{}_RUNNER", target.to_uppercase().replace('-', "_"));
+    cmd.env(runner_env_name, &miri_path);
 
-        if !exit_status.success() {
-            std::process::exit(exit_status.code().unwrap_or(-1))
-        }
+    // Run cargo.
+    if verbose {
+        cmd.env("MIRI_VERBOSE", ""); // this makes `inside_cargo_rustc` verbose.
```
Comment (oli-obk) [inline cargo-miri/bin.rs:444]:
```suggestion
        cmd.env("MIRI_VERBOSE", ""); // this makes `phase_cargo_rustc` verbose.
```

**Review by oli-obk — COMMENTED** [review]
this seems reasonable to me, though it does require quite some hacks (which has never been a blocker for me :laughing: )

**Comment by alecmocatta** [comment]
Getting rid of `cargo metadata` and using `CARGO_TARGET_$TARGET_RUNNER` looks great to me. Eyeballing the changes I believe this fixes the three issues I raised. I'll confirm and share some tests by end of week.

**Comment by alecmocatta** [comment]
I put some tests [here](https://github.com/alecmocatta/cargo-miri-tests).

I observed two issues:

```text
$ MIRIFLAGS='-Zmiri-disable-isolation' cargo miri test
   Compiling root v0.1.0 (/Users/alecmocatta/Documents/deploy/miri-tests)
    Finished test [unoptimized + debuginfo] target(s) in 0.10s
     Running target/x86_64-apple-darwin/debug/deps/root-2b5386c0b2f28133

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running target/x86_64-apple-darwin/debug/deps/harness_false-420a48ffadc82089
fatal error: `cargo-miri` called with unexpected first argument `/Users/alecmocatta/Documents/deploy/miri-tests/target/x86_64-apple-darwin/debug/deps/harness_false-420a48ffadc82089`; please only invoke this binary through `cargo miri`
error: test failed, to rerun pass '--test harness_false'
```
and
```text
$ MIRIFLAGS='-Zmiri-disable-isolation' cargo miri test -p sub
   Compiling sub v0.1.0 (/Users/alecmocatta/Documents/deploy/miri-tests/sub)
    Finished test [unoptimized + debuginfo] target(s) in 0.00s
     Running target/x86_64-apple-darwin/debug/deps/sub-04947ff4d5c37221
error: couldn't read sub/src/main.rs: No such file or directory (os error 2)

error: aborting due to previous error

error: test failed, to rerun pass '-p sub --bin sub'
```

but #1512 and #1514 are fixed.

**Comment by RalfJung** [comment]
@alecmocatta thanks for testing! I also had just noticed the `harness = false` issue, that should be fixed now. I'll check the workspace example.

**Comment by RalfJung** [comment]
Ah, the workspace thing is annoying... the working directory is different between build-time and run-time. Miri now runs in the working directory the test runs, but then the relative path `sub/src/main.rs` makes no sense any more.

**Comment by RalfJung** [comment]
@alecmocatta all right, this should work now with the latest version.

**Comment by RalfJung** [comment]
@oli-obk All boxes are checked, this PR is ready for review now. :)

**Comment by RalfJung** [comment]
And now CI is happy, too. :)

**Comment by RalfJung** [comment]
Hm, let's make sure it also works on Windows hosts.
@bors try

**Comment by alecmocatta** [comment]
Great work, thanks @RalfJung! Confirmed fixes all the issues I've observed.

**Comment by ehuss** [comment]
I was a little unclear, why not use `cargo check` instead of run/test?  I can guess for `cargo run`, it is not easy to determine which binary to run (interpreting `default-run` and such), but is that all?  It seems like you could avoid some of the `link` hacks if it used `cargo check` directly.

**Comment by RalfJung** [comment]
> I was a little unclear, why not use `cargo check` instead of run/test? I can guess for `cargo run`, it is not easy to determine which binary to run (interpreting `default-run` and such), but is that all? It seems like you could avoid some of the `link` hacks if it used `cargo check` directly.

That's what we did before. For `cargo test` it is completely out of the question as we'd have to reimplement all the logics to discover test targets. See the long list of issues this PR fixes for why that just doesn't fly.

So given that we have to use original `cargo test` (and thus need to implement the link hacks), I figured we might just do the same with `cargo run` to keep things consistent.

**Comment by RalfJung** [comment]
> cargo check --tests --examples I think is pretty equivalent to cargo test (doesn't include doctests, not sure if miri handles those).

But then how do we emulate the behavior of `cargo test --lib` and similar filter options?

Miri does not handle doctests yet but that is certainly planned for the future.

**Comment by ehuss** [comment]
> But then how do we emulate the behavior of `cargo test --lib` and similar filter options? And what about `cargo test --workspace`?

`cargo check --profile=test --lib` and `cargo check --profile=test --tests --examples --workspace`.  In other words, if no target flags are passed, default to `--tests --examples`, and if there are any passed, then just pass them through. The package and target filter options are the same on all commands.  

I realize that requires a little more work to check flags and such, just tossing it out as another option if the `--emit` hacking is too much trouble. 

> Miri does not handle doctests yet but that is certainly planned for the future.

If it were to use `cargo check`, doctests would need to be done somewhat manually.  That is, run `cargo check --lib --profile=test`, and for the lib execution, `cargo miri` would need to both generate the `rmeta`, and then manually run `rustdoc` with the `--test-builder` option set to `cargo-miri` (I think). That would require some work to build the `rustdoc` flags, since there is no equivalent of `cargo check` for doctests (thinking about that more, I wonder if rustdoc even works with rmeta files. Seems like it should, but I doubt anyone has ever tried.)

**Comment by RalfJung** [comment]
> I realize that requires a little more work to check flags and such, just tossing it out as another option if the --emit hacking is too much trouble.

Interesting, `cargo check` is much more versatile than I thought.

For now I think I'd like to go with the approach proposed here (unless @oli-obk objects), but I will leave a comment with what you proposed. What this approach here handles better than `cargo check` is actually running the program -- with `cargo check` we'd have to collect somewhere which binaries were compiled and iterate them ourselves, and then we still might have to manually set some environment variables. Overwriting the `RUNNER` means we can rely on cargo also for enumerating the binaries and running them in the right environment.

**Comment by RalfJung** [comment]
(fixed conflict)

**Comment by oli-obk** [comment]
@bors r+

nope I'm fine continuing with the current approach

**Pull request merged**